services/elastic_ip_manager/src/main/java/com/futurewei/alcor/elasticipmanager/controller/ElasticIpController.java

```diff
@@ -0,0 +1,412 @@
+                throw new ElasticIpPipVersionException();
+            }
+
+            if (elasticIpInfo.getPrivateIp() != null) {
+                if (elasticIpInfo.getPrivateIp() == null) {
+                    elasticIpInfo.setPrivateIpVersion(IpVersion.IPV4.getVersion());
+                }
+                if (this.isIpAddressInvalid(elasticIpInfo.getPrivateIp(), elasticIpInfo.getPrivateIpVersion())) {
+                    throw new ElasticIpPipAddressException();
+                }
+            }
+        }
+    }
+
+    private void createElasticIpRangeParameterProcess(ElasticIpRangeInfo elasticIpRangeInfo) throws Exception {
+
+        if (elasticIpRangeInfo == null) {
+            throw new ElasticIpQueryFormatException();
+        }
+
+        if (elasticIpRangeInfo.getIpVersion() == null) {
+            elasticIpRangeInfo.setIpVersion(IpVersion.IPV4.getVersion());
+        } else if (this.isIpVersionInvalid(elasticIpRangeInfo.getIpVersion())) {
+            throw new ElasticIpRangeVersionException();
+        }
+
+        List<ElasticIpRange.AllocationRange> allocationRanges = elasticIpRangeInfo.getAllocationRanges();
+        if (allocationRanges != null) {
+            if (this.isAllocationRangesInvalid(elasticIpRangeInfo.getIpVersion(), allocationRanges)) {
+                throw new ElasticIpRangeBadRangesException();
+            }
+        } else {
+            elasticIpRangeInfo.setAllocationRanges(new ArrayList<>());
+        }
+
+        if (elasticIpRangeInfo.getName() == null) {
+            elasticIpRangeInfo.setName("");
+        }
+
+        if (elasticIpRangeInfo.getDescription() == null) {
+            elasticIpRangeInfo.setDescription("");
+        }
+    }
+
+    private void updateElasticIpRangeParameterProcess(String elasticipRangeId,
+                                                      ElasticIpRangeInfo elasticIpRangeInfo) throws Exception {
+
+        if (elasticIpRangeInfo == null) {
+            throw new ElasticIpQueryFormatException();
+        }
+
+        if (StringUtils.isEmpty(elasticipRangeId)) {
+            throw new ElasticIpRangeNoIdException();
+        } else if (elasticIpRangeInfo.getId() == null) {
+            elasticIpRangeInfo.setId(elasticipRangeId);
+        } else if (!elasticipRangeId.equals(elasticIpRangeInfo.getId())) {
+            throw new ElasticIpIdConfilictException();
+        }
+
+        if (elasticIpRangeInfo.getIpVersion() != null && this.isIpVersionInvalid(elasticIpRangeInfo.getIpVersion())) {
+            throw new ElasticIpRangeVersionException();
+        }
+
+        List<ElasticIpRange.AllocationRange> allocationRanges = elasticIpRangeInfo.getAllocationRanges();
+        if (allocationRanges != null) {
+            if (this.isAllocationRangesInvalid(elasticIpRangeInfo.getIpVersion(), allocationRanges)) {
+                throw new ElasticIpRangeBadRangesException();
+            }
+        }
+    }
+
+    @PostMapping("/project/{project_id}/elasticips")
```

Comment:
Recommend to add comments to public controller methods as well as important private methods.

Reply:
Yes, I will do that

Reply:
Thank you! Those comments are all good.